File: apps/desktop/src/features/notepad/intentParser.ts
/**
 * Deterministic, local-only line classifier for Notepad Intents.
 *
 * No LLM is used for detection. This is the single entry point that decides
 * whether a line of note text becomes a Notepad Intent.
 *
 * Rules (slice 1):
 *   - Plain text (no @, no TODO prefix)         -> NOTE
 *   - # TODO / - TODO / TODO / [ ] / [x]        -> TODO
 *   - /word                                      -> COMMAND (reserved; no-op in slice 1)
 *   - @<token> ... with token = a-z[a-z0-9_-]{1,32}
 *       - if token is a registered capability:
 *           - enabled      -> TASK (or specialized type by verb)
 *           - disabled     -> TASK with status=DEFERRED
 *       - else              -> NOTE (silently treated as plain text)
 */

import { getCapability, isExecutable } from "./capabilities";
import type { RiskLevel, Capability } from "./capabilities";

export type IntentType =
  | "NOTE"
  | "TODO"
  | "TASK"
  | "COMMAND"
  | "RESEARCH_REQUEST"
  | "DRAFT_REQUEST"
  | "AUTOMATION_REQUEST"
  | "EXTERNAL_ACTION"
  | "MULTI_ACTION_WORKFLOW";

export type IntentStatus =
  | "DETECTED"
  | "PENDING_APPROVAL"
  | "APPROVED"
  | "REJECTED"
  | "SKIPPED"
  | "ROUTED"
  | "RUNNING"
  | "COMPLETED"
  | "FAILED"
  | "CANCELLED"
  | "DEFERRED";

export interface DetectedIntent {
  id: string;
  note_id: string;
  line_number: number;
  raw_text: string;
  type: IntentType;
  entities: Record<string, string>;
  capability_id: string | null;
  requested_action: string;
  risk: RiskLevel;
  approval_required: boolean;
  confidence: number;
  status: IntentStatus;
  created_at: string;
  updated_at: string;
  result: null;
  failure: null;
  task_id: null;
  confirmation_id: null;
}

function atTokenMatch(line: string): RegExpMatchArray | null {
  // Anchor to start-of-line first so leading @ai works.
  const start = line.match(/^@([a-z][a-z0-9_-]{1,32})\b(.*)$/);
  if (start) return start;
  // Otherwise look for an @-token preceded by whitespace.
  const m = line.match(/(?<=\s)@([a-z][a-z0-9_-]{1,32})\b(.*)$/);
  return m;
}
const TODO_PATTERN = /^\s*(?:[-*]\s*)?(?:#\s*)?(?:\[(?:\s|x)?\]\s*)?TODO\b[:\s]*(.*)$/i;
const COMMAND_PATTERN = /^\/([a-z][a-z0-9_-]{0,32})\b(.*)$/;

function uuid(): string {
  // Simple, dependency-free UUIDv4-ish. Sufficient for client-side intent ids.
  const rnd = () => Math.floor(Math.random() * 0x10000).toString(16).padStart(4, "0");
  return `${rnd()}${rnd()}-${rnd()}-4${rnd().slice(1)}-${(8 + Math.floor(Math.random() * 4)).toString(16)}${rnd().slice(1)}-${rnd()}${rnd()}${rnd()}`;
}

function nowIso(): string {
  return new Date().toISOString();
}

function inferVerbForCapability(cap: Capability, rest: string): string {
  // The first word after @cap is treated as the verb when supported,
  // otherwise "summarize" is the safe default for @ai.
  const firstWord = (rest.trim().split(/\s+/)[0] ?? "").toLowerCase();
  if (cap.supportedActions.includes(firstWord)) {
    return firstWord;
  }
  return cap.supportedActions[0] ?? "";
}

function classifyRisk(capabilityId: string, verb: string, text: string): { risk: RiskLevel; approval: boolean } {
  const lower = (text || "").toLowerCase();
  const hasInjection =
    lower.includes("ignore previous") ||
    lower.includes("ignore all previous") ||
    lower.includes("system prompt") ||
    lower.includes("disregard the above");
  if (hasInjection) {
    return { risk: "HIGH", approval: true };
  }
  if (capabilityId === "ai") {
    if (verb === "research") return { risk: "MEDIUM", approval: true };
    if (verb === "summarize" || verb === "draft" || verb === "rewrite" || verb === "extract") {
      return { risk: "LOW", approval: false };
    }
    return { risk: "LOW", approval: false };
  }
  return { risk: "MEDIUM", approval: true };
}

function typeForVerb(capabilityId: string, verb: string): IntentType {
  if (capabilityId !== "ai") {
    return "EXTERNAL_ACTION";
  }
  switch (verb) {
    case "research":
      return "RESEARCH_REQUEST";
    case "draft":
      return "DRAFT_REQUEST";
    case "summarize":
    case "rewrite":
    case "extract":
      return "TASK";
    default:
      return "TASK";
  }
}

export function detectIntentsForLine(
  rawLine: string,
  lineNumber: number,
  noteId: string
): DetectedIntent | null {
  const trimmed = rawLine.replace(/\r$/, "");
  if (trimmed.trim().length === 0) return null;

  // TODO classification first (so a TODO line containing @ is still a TODO
  // only if it begins with TODO — the AT matcher below will catch @-first
  // lines and prefer them, per spec R10).
  const todoMatch = trimmed.match(TODO_PATTERN);
  if (todoMatch) {
    return {
      id: uuid(),
      note_id: noteId,
      line_number: lineNumber,
      raw_text: trimmed,
      type: "TODO",
      entities: { rest: todoMatch[1] ?? "" },
      capability_id: null,
      requested_action: "",
      risk: "LOW",
      approval_required: false,
      confidence: 1.0,
      status: "DETECTED",
      created_at: nowIso(),
      updated_at: nowIso(),
      result: null,
      failure: null,
      task_id: null,
      confirmation_id: null,
    };
  }

  // Command (reserved; safe no-op in slice 1).
  if (COMMAND_PATTERN.test(trimmed)) {
    return {
      id: uuid(),
      note_id: noteId,
      line_number: lineNumber,
      raw_text: trimmed,
      type: "COMMAND",
      entities: {},
      capability_id: null,
      requested_action: "",
      risk: "LOW",
      approval_required: false,
      confidence: 1.0,
      status: "SKIPPED",
      created_at: nowIso(),
      updated_at: nowIso(),
      result: null,
      failure: null,
      task_id: null,
      confirmation_id: null,
    };
  }

  // @-capability classification.
  const atMatch = atTokenMatch(trimmed);
  if (atMatch) {
    const token = atMatch[1].toLowerCase();
    const rest = atMatch[2] ?? "";
    const cap = getCapability(token);
    if (cap) {
      const verb = inferVerbForCapability(cap, rest);
      const { risk, approval } = classifyRisk(cap.id, verb, trimmed);
      const executable = isExecutable(cap.id);
      return {
        id: uuid(),
        note_id: noteId,
        line_number: lineNumber,
        raw_text: trimmed,
        type: typeForVerb(cap.id, verb),
        entities: { verb, rest: rest.trim() },
        capability_id: cap.id,
        requested_action: verb,
        risk,
        approval_required: approval,
        confidence: executable ? 1.0 : 0.0,
        status: executable ? "DETECTED" : "DEFERRED",
        created_at: nowIso(),
        updated_at: nowIso(),
        result: null,
        failure: null,
        task_id: null,
        confirmation_id: null,
      };
    }
    // Unknown @capability: fall through and treat as plain NOTE.
  }

  return null;
}

export function detectIntentsInNote(text: string, noteId: string): DetectedIntent[] {
  if (!text) return [];
  const lines = text.split("\n");
  const out: DetectedIntent[] = [];
  for (let i = 0; i < lines.length; i++) {
    const intent = detectIntentsForLine(lines[i], i + 1, noteId);
    if (intent) out.push(intent);
  }
  return out;
}

File: apps/backend/app/notepad/intent_parser.py
"""Server-side intent parser (Slice 1).

Mirrors apps/desktop/src/features/notepad/intentParser.ts. Detection is
deterministic and local; no LLM is used. The parser produces an IntentDTO-
shaped dict so the route can return it directly.
"""
from __future__ import annotations

import re
import uuid
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from app.notepad.capabilities import CAPABILITIES, get_capability, is_executable
from app.notepad.schemas import (
    IntentType,
    RiskLevel,
)


_AT_TOKEN_PATTERN = re.compile(r"(?:^|\s)@([a-z][a-z0-9_-]{1,32})\b(.*)$")

def _at_token_match(line: str):
    """Match an @-capability token anywhere on the line. Returns the match
    object or None. We use search+lookbehind so the prefix whitespace is
    not consumed and we can still anchor to start-of-line."""
    # Try start-of-line first (so leading @ai still works).
    m = re.match(r"@([a-z][a-z0-9_-]{1,32})\b(.*)$", line)
    if m:
        return m
    # Otherwise find an @-token preceded by whitespace.
    return re.search(r"(?<=\s)@([a-z][a-z0-9_-]{1,32})\b(.*)$", line)
_TODO_PATTERN = re.compile(r"^\s*(?:[-*]\s*)?(?:#\s*)?(?:\[(?:\s|x)?\]\s*)?TODO\b[:\s]*(.*)$", re.IGNORECASE)
_COMMAND_PATTERN = re.compile(r"^/([a-z][a-z0-9_-]{0,32})\b(.*)$")

_INJECTION_INDICATORS = (
    "ignore previous",
    "ignore all previous",
    "system prompt",
    "disregard the above",
)


def _utc_now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _new_id() -> str:
    return uuid.uuid4().hex


def _infer_verb(cap_actions: tuple, rest: str) -> str:
    first = (rest.strip().split() or [""])[0].lower()
    if first in cap_actions:
        return first
    return cap_actions[0] if cap_actions else ""


def _classify_risk(capability_id: str, verb: str, text: str) -> Dict[str, Any]:
    lower = (text or "").lower()
    if any(ind in lower for ind in _INJECTION_INDICATORS):
        return {"risk": RiskLevel.HIGH, "approval": True}
    if capability_id == "ai":
        if verb == "research":
            return {"risk": RiskLevel.MEDIUM, "approval": True}
        if verb in ("summarize", "draft", "rewrite", "extract"):
            return {"risk": RiskLevel.LOW, "approval": False}
        return {"risk": RiskLevel.LOW, "approval": False}
    return {"risk": RiskLevel.MEDIUM, "approval": True}


def _type_for_verb(capability_id: str, verb: str) -> IntentType:
    if capability_id != "ai":
        return IntentType.EXTERNAL_ACTION
    mapping = {
        "research": IntentType.RESEARCH_REQUEST,
        "draft": IntentType.DRAFT_REQUEST,
        "summarize": IntentType.TASK,
        "rewrite": IntentType.TASK,
        "extract": IntentType.TASK,
    }
    return mapping.get(verb, IntentType.TASK)


def detect_intents_for_line(raw_line: str, line_number: int, note_id: str) -> Optional[Dict[str, Any]]:
    """Return a dict matching IntentDTO fields, or None for plain text."""
    if raw_line is None:
        return None
    trimmed = raw_line.rstrip("\r")
    if not trimmed.strip():
        return None

    # TODO
    todo = _TODO_PATTERN.match(trimmed)
    if todo:
        return {
            "id": _new_id(),
            "note_id": note_id,
            "line_number": line_number,
            "raw_text": trimmed,
            "type": IntentType.TODO.value,
            "entities": {"rest": todo.group(1) or ""},
            "capability_id": None,
            "requested_action": "",
            "risk": RiskLevel.LOW.value,
            "approval_required": False,
            "confidence": 1.0,
            "status": "DETECTED",
            "task_id": None,
            "confirmation_id": None,
            "result": None,
            "failure": None,
            "created_at": _utc_now(),
            "updated_at": _utc_now(),
        }

    # COMMAND (reserved; safe no-op in slice 1)
    if _COMMAND_PATTERN.match(trimmed):
        return {
            "id": _new_id(),
            "note_id": note_id,
            "line_number": line_number,
            "raw_text": trimmed,
            "type": IntentType.COMMAND.value,
            "entities": {},
            "capability_id": None,
            "requested_action": "",
            "risk": RiskLevel.LOW.value,
            "approval_required": False,
            "confidence": 1.0,
            "status": "SKIPPED",
            "task_id": None,
            "confirmation_id": None,
            "result": None,
            "failure": None,
            "created_at": _utc_now(),
            "updated_at": _utc_now(),
        }

    # @-capability
    at = _at_token_match(trimmed)
    if at:
        token = at.group(1).lower()
        rest = at.group(2) or ""
        cap = get_capability(token)
        if cap is not None:
            verb = _infer_verb(cap.supported_actions, rest)
            risk = _classify_risk(cap.id, verb, trimmed)
            executable = is_executable(cap.id)
            return {
                "id": _new_id(),
                "note_id": note_id,
                "line_number": line_number,
                "raw_text": trimmed,
                "type": _type_for_verb(cap.id, verb).value,
                "entities": {"verb": verb, "rest": rest.strip()},
                "capability_id": cap.id,
                "requested_action": verb,
                "risk": risk["risk"].value if hasattr(risk["risk"], "value") else risk["risk"],
                "approval_required": risk["approval"],
                "confidence": 1.0 if executable else 0.0,
                "status": "DETECTED" if executable else "DEFERRED",
                "task_id": None,
                "confirmation_id": None,
                "result": None,
                "failure": None,
                "created_at": _utc_now(),
                "updated_at": _utc_now(),
            }
        # Unknown @capability: fall through to plain text.

    return None


def detect_intents_in_note(text: str, note_id: str) -> List[Dict[str, Any]]:
    if not text:
        return []
    out: List[Dict[str, Any]] = []
    for idx, line in enumerate(text.split("\n")):
        detected = detect_intents_for_line(line, idx + 1, note_id)
        if detected is not None:
            out.append(detected)
    return out
